scan/scan_common.go

```diff
@@ -132,39 +132,36 @@ type ScannerResult struct {
 // FamilyResult contains a scan response for a single Family
 type FamilyResult map[string]ScannerResult
 
-// RunScans iterates over AllScans, running scans matching the family and scanner
-// regular expressions.
-func (fs FamilySet) RunScans(host, family, scanner string, dur time.Duration) (map[string]FamilyResult, error) {
-	if _, _, err := net.SplitHostPort(host); err != nil {
-		host = net.JoinHostPort(host, "443")
-	}
-
-	familyRegexp, err := regexp.Compile(family)
-	if err != nil {
-		return nil, err
-	}
+// RunScans iterates over AllScans, running each scan that matches the family
+// and scanner regular expressions concurrently.
+func (fs FamilySet) RunScans(host, family, scanner string) (chan map[string]FamilyResult, chan error) {
+	results := make(chan map[string]FamilyResult)
+	errors := make(chan error)
 
-	scannerRegexp, err := regexp.Compile(scanner)
-	if err != nil {
-		return nil, err
-	}
+	go func() {
+		if _, _, err := net.SplitHostPort(host); err != nil {
+			host = net.JoinHostPort(host, "443")
+		}
 
-	results := make(chan map[string]FamilyResult)
-	timeout := make(chan bool)
+		familyRegexp, err := regexp.Compile(family)
+		if err != nil {
+			return
+		}
 
-	if dur > 0 {
-		go startTimer(dur, timeout)
-	}
+		scannerRegexp, err := regexp.Compile(scanner)
+		if err != nil {
+			return
+		}
 
-	familyResults := make(map[string]FamilyResult)
+		familyResults := make(map[string]FamilyResult)
 
-	go func() {
 		for familyName, family := range fs {
 			if familyRegexp.MatchString(familyName) {
 				scannerResults := make(map[string]ScannerResult)
 
 				for scannerName, scanner := range family.Scanners {
 					if scannerRegexp.MatchString(scannerName) {
+
 						grade, output, err := scanner.Scan(host)
```

Comment:
I think the biggest win we can get is by running individual scans (or maybe Families) in parallel and streaming back the results.

It's much more common to be waiting on multiple scans on a single host than scans across multiple hosts (especially from API calls).